Performance tuning

VincentKueszter opened the pull request:
Hello,

while searching for some method to generate folds and wrinkles in clothing I found this add-on and I liked it a lot! The only problem (for me) is that it's not running real-time, so I tried to have a look at the source code, to see if there is any way to optimize it.

My first approach was to try and write the squeeze/stretch data to a texture, but that was even slower (apparently the API is the limiting factor here https://devtalk.blender.org/t/bpy-data-images-perf-issues/6459/8). I dropped that and tried to improve the current approach instead.

This PR now contains all "small things" I managed add to squeeze the last bit of performance out of the script:
- pre-calculate original edge lengths
- generate vertex color list to avoid having to read from vertex group
- added option to disable vertex group calculation
- early return if vertex group calculation is disabled
- made "Enable Vertex Groups" disabled by default, to avoid having to delete them after activating the script (if you don't want to use them).

I am not really proficient in Python or the Blender API, so if there is anything that I could improve before merging, please don't hesitate to mention it!

Kind regards,
Vince

VincentKueszter committed:
- added option to disable vertex group calculation
- early return if vertex group calculation is disabled
- pre-calculate original edge lengths
- generate vertex color list to avoid having to read from vertex group

VincentKueszter committed:
- updated readme

VincentKueszter committed:
- made "Enable Vertex Groups" disabled by default, to avoid having to delete them after activating the script (if you don't want to use them).

VincentKueszter committed:
- improved grammar
- fixed wrong version number
- improved formatting

ScottishCyclops commented on tensionmap.py:
```diff
@@ -147,35 +154,60 @@ def tm_update(obj, context):
-            group_squeeze.add(add_index, obj.data.tm_minimum, "REPLACE")
-            group_stretch.add(add_index, max(obj.data.tm_minimum, min(obj.data.tm_maximum, weights[i])), "REPLACE")
+            stretch_value = max(obj.data.tm_minimum, min(obj.data.tm_maximum, weights[i]))
+            if obj.data.tm_enable_vertex_groups:
```
Since the values `stretch_value` and `squeeze_value` are stored at the scope of `for i in range(num_vertices)`, the test on line `169`, repeated on line `176`, can be extracted to outside the if-else, to avoid code repetition.

It also means using:
```python
group_squeeze.add(add_index, squeeze_value, "REPLACE")
group_stretch.add(add_index, stretch_value, "REPLACE")
```
Twice, instead of using:

```python
group_squeeze.add(add_index, obj.data.tm_minimum, "REPLACE")
group_stretch.add(add_index, stretch_value, "REPLACE")

...

group_squeeze.add(add_index, squeeze_value, "REPLACE")
group_stretch.add(add_index, obj.data.tm_minimum, "REPLACE")
```

Which is also more consistent anyway.

VincentKueszter replied:
Right, that makes it much simpler, I'll change it.

ScottishCyclops commented on tensionmap.py:
```diff
@@ -147,35 +154,60 @@ def tm_update(obj, context):
-
+    
+    #create vertex color list for faster access
+    number_of_tm_channels = 2
```
This is a constant, should be at the top level of this file.

VincentKueszter replied:
Done.

ScottishCyclops commented on tensionmap.py:
```diff
@@ -147,35 +154,60 @@ def tm_update(obj, context):
     object_eval.to_mesh_clear()
-
+    
+    #create vertex color list for faster access
```
Spaces between `#` and the comment.
https://www.python.org/dev/peps/pep-0008/#block-comments

I tried to follow the PEP8 style guide. However, as I now read the specs, I see they want capital letters at the beginning of comments.
I never used that in this file, so don't use it now. But the space would make it consistent with the rest of the file.

VincentKueszter replied:
Done.

ScottishCyclops commented on tensionmap.py:
```diff
@@ -147,35 +154,60 @@ def tm_update(obj, context):
+            for loop_vertex_idx, loop_idx in enumerate(polygon.loop_indices):
+                vertex_color = colors_tension.data[loop_idx]
+                vertex_idx = polygon.vertices[loop_vertex_idx]
+                vertex_color.color[0] = vertex_colors[vertex_idx*number_of_tm_channels]
```
I do believe the `# red` and `# green` comments are useful for anyone looking at the code. You could add them back

VincentKueszter replied:
Oh, they were removed by mistake. Re-added!

ScottishCyclops commented on tensionmap.py:
```diff
@@ -147,35 +154,60 @@ def tm_update(obj, context):
+                vertex_color.color[1] = vertex_colors[vertex_idx*number_of_tm_channels+1]
+
+                        
+def tm_update_original_edges(obj):
```
I see a few problems with this function.

First, I do not see how you can improve performance with it. This function is executed as many times as the `tm_update` function, which used to handle this logic.

Plus, you are now introducing a second loop. The previous implementation was using a single loop to compute the original & the deformed edge length. So I see it as a performance decrease.

And finally, using a global variable for this is ugly.
If you wanted to do it nicely, this function would return the `original_edge_lengths`, and `tm_update` would then take it as a parameter.

VincentKueszter replied:
The performance increase is due to the function being called only by `tm_update_selected`, which is only called when clicking the "Update Tension Map" button. It is not called by `tm_update` each frame.

The global variable is very ugly and also the source of another problem: initialization.
The script keeps throwing `IndexError: list index out of range` until you press the button, because `original_edge_lengths` starts empty!
That's what I get for lazily using a global variable after decades of rightful abstinence! 😃

I'll see how I fix that.

ScottishCyclops replied:
Oh I see.
But while that might increase performance, it comes at the great cost of usability. It means you will need to manually update your meshes when you make changes to them.

Also, since I saw 0% of performance increase when using the Vertex Group, I would suppose that this does not come with any performance improvement.
Otherwise, the Vertex Group generation would also be faster, not just the Vertex Color.

What do you think ?

VincentKueszter replied:
Hm, I see. Having to click that button is not a viable solution, of course.
I suppose there must be a way to detect if an object mesh has changed and I already found some [first possibilities](https://docs.blender.org/api/blender2.8/bpy.types.DepsgraphUpdate.html?highlight=is_updated#bpy.types.DepsgraphUpdate.is_updated_geometry).
I'll explore that.

Disabling the Vertex Group generation brought a little bit of performance for me, but it was only a  increase from 15.5 to 16.5 fps on a model with 2000 vertices. The main advantage in my eyes is that you don't need to have two additional vertex groups "polluting" your list if you don't need them.

VincentKueszter replied:
The depsgraph was a dead end, it seems to be related to the modifier stack. Apparently I would have to hash the base model/vertex data in some way to notice it changed and I guess this would be too costly.

I changed the implementation of the edge calculation so that it still is a global variable, but done via a dictionary, containing each object for which the Tension Map is enabled.
It checks during the update if it has seen the object. If not, it generates the data once.
If the Tension Map is disabled, the edge list gets deleted from the dictionary.
If the geometry of an object changed, the user has to click "Update Tension Map" button.
I guess that is the best solution for now. If you have better suggestions for both problems (global var and automatically noticing geometry changes), I'm all ears. 😃

ScottishCyclops replied:
> Disabling the Vertex Group generation brought a little bit of performance for me, but it was only a increase from 15.5 to 16.5 fps on a model with 2000 vertices. The main advantage in my eyes is that you don't need to have two additional vertex groups "polluting" your list if you don't need them.

`Vertex Group + Vertex Color` VS `only Vertex Color` definitively had performance improvements, as I said in my post about performance.
What I am talking about is `the old code with only Vertex Group` VS `the new code with only Vertex Group`.

In that case, I saw no difference in performance.
What that tells me is, the optimization we are talking about in this thread doesn't seem to actually improve performance, otherwise, the new code would be faster than the old one, even if it's only generating the Vertex Group.

So we should keep the checkbox for disabling the Vertex Group, the "pollution" point is valid. But the optimization using that global var and all seems useless.

VincentKueszter replied:
Strange, the optimization with the global var is one of the main performance improvements!

And, I mean, it makes sense:
If vertex group or color generation is enabled, we need to calculate stretch/squeeze during each frame in both cases.
What the global variable stuff does, is moving the calculation of the original edge lengths out of the update function, so we don't have to do that each frame. It is only evaluated if the user presses the button or if we need it the first time.

On my system, simply replacing line 149
`original_edge_length = original_edge_lengths[i]`
with 
`original_edge_length = original_edge_lengths[i] = (obj.data.vertices[first_vertex].co - 
                                    obj.data.vertices[second_vertex].co).length`
(thus bypassing the improvements and calculating it locally again) raises my fps from 30 to 34, with "Enable Vertex Groups" only.
Please try it on your system, maybe some other changes were influencing the result.

Also you are right, I should make one commit per improvement, that makes testing easier, too.

ScottishCyclops replied:
I'm back. I've made [an addon](https://github.com/ScottishCyclops/viewport-fps) to accurately measure the average FPS.

Average FPS over 600 frames, calculated in my test scene for each versions:

Options | 2 1 1 | 2 1 1 no global | 2 1 0
-- | -- | -- | --
All | 11,37 | 11,09 | 8,87
Group only | 12,79 | 12,4 | 12,67
Color only | 11,68 | 11,2 |  

Comparisons (x vs y, x is how much faster in percent)

2 1 1 vs 2 1 1 no global | 2 1 1 vs 2 1 0 | 2 1 1 no global vs 2 1 0
-- | -- | --
2,52 % | 28,18 % | 25,03 %
3,15 % | 0,95 % | -2,13 %
4,29 % |   |  


> Strange, the optimization with the global var is one of the main performance improvements!

So the global optimization (let's refer to it as the "global optimization" from now on) is definitively not the main performance improvement. Without it, we still get a beautiful >25% overall performance increase.

That said, it is still a bit faster, so it is a good idea to explore.

I also noted that with the new code and without the *global optimization*, only calculating the Vertex Group is a tiny bit slower. What that means is your code is better at calculating the Vertex Color, but for Vertex Group, is adds overhead.

Can't help it, since we create additional arrays and such.

Here is what I propose: we ignore the  *global optimization* for this PR, and merge the rest of the improvements.

Then, you can continue working on a good solution to integrate the *global optimization*, and make a new PR for it.

In the meantime, I saw other things that might improve the perf a little for Vertex Group, so I'll work on that.

VincentKueszter replied:
Okay, I ran it, too and I got similar results:

|            | 211  | 211 no Global | 210  |
|------------|------|---------------|------|
| All        | 20,8 | 19,5          | 11,5 |
| Group Only | 31,6 | 28,6          | 30,3 |
| Color Only | 21,9 | 21,1          |      |

You're right, the influence of the global optimization variant has a smaller influence than I thought.
I will commit the 211 no Global variant I created for that test, so you can simply merge that.

Nice FPS addon by the way!
PS: There is a small typo, the toolbar says "Wiewport".

ScottishCyclops replied:
>  PS: There is a small typo, the toolbar says "Wiewport".

On my way to fix that x)

ScottishCyclops commented on tensionmap.py:
```diff
@@ -129,12 +135,13 @@ def tm_update(obj, context):
     # calculate the new weights
-    for edge in obj.data.edges:
+    global original_edge_lengths
+    for i in range(len(obj.data.edges)):
```
I don't see why using 
```python
for i in range(len(obj.data.edges)):
    edge = obj.data.edges[i]
```

Instead of the old

```python
for edge in obj.data.edges:
```

I don't think it is more performant (I might be wrong ?), but it certainly is worse to the eye.

EDIT: I think you needed the index for line `144`. But that won't be necessary depending on what we decide (see the rest of my comments)

VincentKueszter replied:
Resolved by removing the *global optimization*

ScottishCyclops commented on tensionmap.py:
```diff
@@ -249,6 +282,7 @@ def draw(self, context):
         row1.active = context.object.data.tm_active
         row1.operator("tm.update_selected")
         row1.prop(context.object.data, "tm_enable_vertex_colors", text="Enable Vertex Colors")
+        row1.prop(context.object.data, "tm_enable_vertex_groups", text="Enable Vertex Groups")
```
In Blender, vertex groups appear before vertex color (by default).
So I think the `Enable Vertex Groups` should appear before the `Enable Vertex Colors` in our UI too.

VincentKueszter replied:
Done!

ScottishCyclops reviewed (changes requested):
Thank you so much for your contribution :heart: 

I went ahead and reviewed your code !
If you have any comments, let's discuss it here.

I see a potential big improvement because of the way you are storing the computed data outside the vertex group
Besides, it looks better than putting it in the vertex group then taking it back out. Good job.

I will now download your branch and test the performance on my own.

ScottishCyclops commented:
Woups. Your version keeps throwing
```
Traceback (most recent call last):
  File "/home/scott/.config/blender/2.80/scripts/addons/tensionmap.py", line 231, in tm_update_handler
    tm_update(obj, bpy.context)
  File "/home/scott/.config/blender/2.80/scripts/addons/tensionmap.py", line 144, in tm_update
    original_edge_length = original_edge_lengths[i]
IndexError: list index out of range
```

In the console when running. So one more reason to abandon that idea.

ScottishCyclops commented:
Here are my results for performance:

Using only Vertex Group: No difference in FPS
Using Vertex Group & Vertex Color: ~11-13% More FPS

And finally, using only Vertex Color (wasn't possible before), compared to using Vertex Group & Vertex Color in the PR version: Only ~2,5-3% More FPS.

This test is made on a single polygon heavy mesh, then again on two polygon heavy meshes, to see if the performance scales. It seams to scale pretty linearly.

Once you do make the few changes, I'll try it again.

Anyway, it's a nice boost for the vertex color mode. Good job :+1:

VincentKueszter commented:
Thanks for the review! I'll implement the changes, test it and push a new version.

VincentKueszter committed:
- review fixes
- globally save original_edge_length per object and check if not yet calculated
- delete original_edge_length if TensionMap generation is disabled for this object
- made tm_update_original_edges return the edges
- performance improvement by different color setting method

VincentKueszter commented:
I noticed a surprisingly large performance increase on my system doing the following when setting the colors:
This is the base line, resulting in 16fps (model with 2000 vertices)
```
vertex_color.color[0] = vertex_colors[vertex_idx*number_of_tm_channels]
vertex_color.color[1] = vertex_colors[vertex_idx*number_of_tm_channels+1]
```
Then I noticed that when I delete the vertex colors via GUI, they get regenerated instantly as expected, but the blue channel is suddenly 1. So I re-added setting it to zero. Surprisingly, the fps dropped to 13!
```
vertex_color.color[0] = vertex_colors[vertex_idx*number_of_tm_channels]
vertex_color.color[1] = vertex_colors[vertex_idx*number_of_tm_channels+1]
vertex_color.color[2] = 0.0
```
Then I tried this
```
vertex_color.color = (vertex_colors[vertex_idx*number_of_tm_channels],vertex_colors[vertex_idx*number_of_tm_channels+1],0.0,1.0)
```
And suddenly I had 23fps.
Probably some vector optimizations in the background.

ScottishCyclops reviewed (commented):
> I noticed a surprisingly large performance increase on my system doing the following when setting the colors:
> This is the base line, resulting in 16fps (model with 2000 vertices)
> 
> ```
> vertex_color.color[0] = vertex_colors[vertex_idx*number_of_tm_channels]
> vertex_color.color[1] = vertex_colors[vertex_idx*number_of_tm_channels+1]
> ```
> 
> Then I noticed that when I delete the vertex colors via GUI, they get regenerated instantly as expected, but the blue channel is suddenly 1. So I re-added setting it to zero. Surprisingly, the fps dropped to 13!
> 
> ```
> vertex_color.color[0] = vertex_colors[vertex_idx*number_of_tm_channels]
> vertex_color.color[1] = vertex_colors[vertex_idx*number_of_tm_channels+1]
> vertex_color.color[2] = 0.0
> ```
> 
> Then I tried this
> 
> ```
> vertex_color.color = (vertex_colors[vertex_idx*number_of_tm_channels],vertex_colors[vertex_idx*number_of_tm_channels+1],0.0,1.0)
> ```
> 
> And suddenly I had 23fps.
> Probably some vector optimizations in the background.

Nice work ! Indeed, setting the vector at once would be faster, because accessing Index 0,  then 1, then 2 is slow by design.
That's what performance improvements is all about :+1:

ScottishCyclops commented:
Would have been great to make one commit per review element, because some of your changes are now valid, but in the same commit are things we are still discussing :thinking: 

This is my first "Github Review" experience, so let's just do our best ^^

VincentKueszter commented:
Sure thing, I'll do that next time! 
I am also rather new to the way GitHub does this. In BitBucket you can create tasks, I think we can replicate that via code comments, since I can mark them as resolved.

VincentKueszter committed:
removed the "global optimization"

ScottishCyclops reviewed (approved):
All good. I'll merge it now.
Expect a subsequent commit of me fixing a few more style inconsistencies. It is faster if I fix them myself.
